server/client.go

```diff
@@ -346,20 +346,23 @@ func nbPoolGet(sz int) []byte {
 	}
 }
 
-func nbPoolPut(b []byte) {
-	switch cap(b) {
-	case nbPoolSizeSmall:
-		b := (*[nbPoolSizeSmall]byte)(b[0:nbPoolSizeSmall])
-		nbPoolSmall.Put(b)
-	case nbPoolSizeMedium:
-		b := (*[nbPoolSizeMedium]byte)(b[0:nbPoolSizeMedium])
-		nbPoolMedium.Put(b)
-	case nbPoolSizeLarge:
-		b := (*[nbPoolSizeLarge]byte)(b[0:nbPoolSizeLarge])
-		nbPoolLarge.Put(b)
-	default:
-		// Ignore frames that are the wrong size, this might happen
-		// with WebSocket/MQTT messages as they are framed
+func nbPoolPut(in []byte) {
+	ca := cap(in)
+	for in = in[:ca]; ca >= nbPoolSizeSmall; ca = cap(in) {
```

Comment:
This is hard to follow. What is this doing?

Reply:
At this point the input `in` length might not match the capacity, so we size the slice back up to the full capacity. Then the loop will start checking if the remaining capacity is at least as big as one of the pool sizes and will return 

The idea being that if we've used `append` a lot and the buffer capacity is no longer one of our three standard sizes, like the compressor will with the `bytes.Buffer`, that we can still reclaim the underlying arrays afterwards by chunking it back down into one of our three sizes.

Reply:
I am concerned that might cause GC problems with interior pointers and causing it not to release anything.

Reply:
I've reverted the changes to `nbPoolPut` for now, it seems that we don't generate that many slices larger than the largest pool size so we maintain most of the performance increase.